### PR TITLE
added types for styletron-engine-monolithic

### PR DESCRIPTION
**Pull request by tmgldn**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/styletron-engine-monolithic/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
+// Type definitions for styletron-engine-monolithic 0.0
+// Project: https://github.com/styletron/styletron
+// Definitions by: Tom Golden <https://github.com/tbjgolden>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.9
```
Comment (peterblazejewicz) [inline types/styletron-engine-monolithic/index.d.ts:5]:
```suggestion
```
https://github.com/DefinitelyTyped/DefinitelyTyped#support-window

File: types/styletron-engine-monolithic/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+import { KeyframesObject, FontFace as FontFaceObject, StandardEngine, StyleObject } from 'styletron-standard';
+
+export class SequentialIDGenerator {
```
Comment (peterblazejewicz) [inline types/styletron-engine-monolithic/index.d.ts:8]: those are not exported symbols, how they could be here?
https://github.com/styletron/styletron/blob/master/packages/styletron-engine-monolithic/src/index.js#L3-L4

Comment (tbjgolden) [reply types/styletron-engine-monolithic/index.d.ts:8]: ah - my bad; should have used declare here, 1 sec

Comment (peterblazejewicz) [reply types/styletron-engine-monolithic/index.d.ts:8]: that''s OK, you can eventually export types (interface/type) if that is somehow useful when consuming this package.

File: types/styletron-engine-monolithic/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    idGenerator: SequentialIDGenerator;
+    onNewCache: (key: string, cache: Cache<T>, insertAtIndex: number) => any;
+    onNewValue: (cache: Cache<T>, id: string, value: T) => any;
+    sortedCacheKeys: string[];
+    constructor(idGenerator: SequentialIDGenerator, onNewCache: () => any, onNewValue: () => any);
+    getCache(key: string): Cache<T>;
+    getSortedCacheKeys(): string[];
+}
+
+declare type hydrateType = HTMLCollectionOf<HTMLStyleElement> | HTMLStyleElement[] | NodeListOf<HTMLStyleElement>;
```
Comment (peterblazejewicz) [inline types/styletron-engine-monolithic/index.d.ts:38]:
```suggestion
declare type HydrateType = HTMLCollectionOf<HTMLStyleElement> | HTMLStyleElement[] | NodeListOf<HTMLStyleElement>;
```
please amned this elsewhere if used, stick to JS/TS naming convention (unless there is obvious reason to not to)

File: types/styletron-engine-monolithic/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+declare interface Sheet {
+    css: string;
+    attrs: { [key: string]: string };
+}
+
+export class Client implements StandardEngine {
+    constructor();
```
Comment (peterblazejewicz) [inline types/styletron-engine-monolithic/index.d.ts:45]:
```suggestion
```

File: types/styletron-engine-monolithic/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    keyframesSheet: HTMLStyleElement;
+    styleCache: MultiCache<{ pseudo: string; block: string }>;
+    keyframesCache: Cache<KeyframesObject>;
+    fontFaceCache: Cache<FontFaceObject>;
+    renderStyle(style: StyleObject): string;
+    renderKeyframes(keyframes: KeyframesObject): string;
+    renderFontFace(fontFace: FontFaceObject): string;
+}
+export class Server implements StandardEngine {
+    constructor();
```
Comment (peterblazejewicz) [inline types/styletron-engine-monolithic/index.d.ts:57]:
```suggestion
```

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@tbjgolden thx!

**Comment by tbjgolden** [comment]
Ready to merge

**Pull request merged**